### ISSUE TITLE
Error in the tests

### ISSUE THREAD

**Issue by LeoLai930603** [Kaffaljidhmah2/DeepLearningFramework#1]
I think this line will not get passed, because there is an assertion in `functional::matmul` which requires the first operand has `dim == 2`

https://github.com/Kaffaljidhmah2/DeepLearningFramework/blob/a6248bed313164063785657fcf259eca2cc43d0a/src/test.cpp#L170

**Comment by Kaffaljidhmah2** [comment]
Yeah, you are right. There is a lot of room for improvement. I stopped to maintain this repo after I could use it to do MNIST.
